new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,224 @@
+    window.history.pushState({}, "", newUrl);
+};
+
+/**
+ *
+ */
+const getAllWindowSearchParams = () => {
+    const searchParams = new URLSearchParams(window.location.search.substring(1));
+    const urlSearchParams: UrlParamsType = {};
+
+    // TODO: use Ajv to read and validate
+    const filePath = searchParams.get(SEARCH_PARAM_NAME.FILE_PATH);
+    if (null !== filePath) {
+        urlSearchParams[SEARCH_PARAM_NAME.FILE_PATH] = filePath;
+    }
+
+    return urlSearchParams;
+};
+
+/**
+ *
+ */
+const getAllWindowHashParams = () => {
+    const hashParams = new URLSearchParams(window.location.hash.substring(1));
+    const urlHashParams: UrlHashParams = {};
+
+    // TODO: use Ajv to read and validate
+    const logEventNum = hashParams.get(HASH_PARAM_NAME.LOG_EVENT_NUM);
+    if (null !== logEventNum) {
+        urlHashParams[HASH_PARAM_NAME.LOG_EVENT_NUM] = Number(logEventNum);
+    }
+
+    return urlHashParams;
+};
+
+
+/**
+ * Copies the current window's URL to the clipboard. If any `updates` parameters are specified,
+ * the copied URL will include these modifications, while the original window's URL remains
+ * unchanged.
+ *
+ * @param searchParamUpdates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ * @param hashParamsUpdates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ */
+const copyToClipboard = (
+    searchParamUpdates: UrlSearchParamUpdatesType,
+    hashParamsUpdates: UrlHashParamUpdatesType,
+) => {
+    const newUrl = new URL(window.location.href);
+    newUrl.search = getUpdatedSearchParams(searchParamUpdates).toString();
+    newUrl.hash = getUpdatedHashParams(hashParamsUpdates).toString();
+    navigator.clipboard.writeText(newUrl.toString())
+        .then(() => {
+            console.log("URL copied to clipboard.");
+        })
+        .catch((error: unknown) => {
+            console.error("Failed to copy URL to clipboard:", error);
+        });
+};
+
+/**
+ * Provides a context for managing URL parameters and hash values,
+ * including utilities for setting search and hash parameters,
+ * and copying the current URL with these parameters to the clipboard.
+ *
+ * @param props
+ * @param props.children The child components that will have access to the context.
+ * @return
+ */
+const UrlContextProvider = ({children}: UrlContextProviderProps) => {
+    const [urlParams, setUrlParams] = useState<UrlParamsType>({});
+
+    useEffect(() => {
+        setUrlParams({
+            ...getAllWindowSearchParams(),
+            ...getAllWindowHashParams(),
+        });
+
+        const handleHashChange = () => {
+            // FIXME: handle removal of hash params
+            setUrlParams((v) => ({
+                ...v,
+                ...getAllWindowHashParams(),
+            }));
```

Comment:
I think we can cache the search parameters with help of `useMemo` or `useRef`. Then we can write this as 
```suggestion
            setUrlParams((v) => ({
                ...cachedAllWindowSearchParams, // or any other shorter name
                ...getAllWindowHashParams(),
            }));
```